Support for Espressif chipsets via ESP-IDF

vikramdattu opened the pull request:
*** This is a placeholder PR: Not to be merged ****

This PR is meant to showcase the work which will bring the support for ESP-IDF build support and hence the Espressif chipsets.

## Structure:
 - Changes specific to ESP-IDF are put under `esp_port` directory.
 - Go to the example [webrtc_classic](esp_port/examples/webrtc_classic) and build it as ESP-IDF example.

## Setup ESP-IDF 
Setup ESP_IDF with release/v5.4 branch: https://docs.espressif.com/projects/esp-idf/en/stable/esp32/get-started/linux-macos-setup.html

## BUILD
 ```
cd esp_port/examples/webrtc_classic
idf.py set-target esp32s3 # Or esp32p4 etc
idf.py build
 ```
 
 ### Flash the code on chip and run monitor
 ```
 idf.py flash
 idf.py monitor
 ```
 
The work is under progress, will keep rebasing as and when common functionality code keeps getting into main repository.

@sirknightj @unicornss 

We have two options:
1. Upstream and merge all the changes to this repo: Not as-is of course and will have reviews from structure and code point of view
2. The `esp_port` maintained by Espressif with some suitable name under espressif/ organisation. It will use `amazon-kinesis-video-sdk-c` as a submodule or need developer to set path of cloned repo.

We can keep this PR for discussion ground continue discussion on other aspects of how best we can go about this. I prefer the option 1, simply because it remains maintained under single roof without needing to constant rebasing.

vikramdattu committed:
Increase socket connection resilience to memory errors

 - Retries increased from 3 to 10
 - Consider ENOMEM or PBUF errors also for retries

These changes help to retain sockets longer for embedded devices

vikramdattu committed:
Timeout increase tests for ESP

vikramdattu committed:
Use custom thread implementation for ESP platforms to save memory

 - Use custom thread implementation for receiveLwsMessageWrapper
 - Custom thread implementation for connectionListener

vikramdattu committed:
Network.c: Changes to support ESP-IDF

 - Added ifaddr altenate for ESP-IDF
 - Created abstracted API for IDF_CMAKE platform to query local IP
 - Do not treat SO_SNDBUF not available as fatal

vikramdattu committed:
ESP32 specific change for memory fix

vikramdattu committed:
ESP port: Added needed components and examples

 - Cleanly added the support under esp_port directory
 - Added ESP specific components, submodules and an webrtc_classic example
 - Added `app_common` needed as a common functionality for webrtc example/s.
 - Added network_adapter example: Needed when we want to use the webrtc_classic on P4+C6 board

vikramdattu committed:
Added WebRTC split functionality

 - Added streaming_only and signaling_only examples
 - streaming_only can be flashed on P4 and signaling only on C6
 - The message exchange between the two happens over webrtc_bridge

vikramdattu committed:
Prefer dynamic allocations over large STATIC arrays

 - Use dynamically allocated WSS URL
 - Use dynamic allocations for singaling payload
 - Use required only allocations in LwsApiCalls instead large static arrays

vikramdattu committed:
ESP port: Use esp_websocket_client as an alternate implementation

 - Added esp_websocket_client over libwebsockets for ESP-IDF
 - Using this as default for ESP port as it provides better flexibility for ESP-IDF

vikramdattu committed:
webrtc message utils

 - Refactored webrtc_message re-assembly logic to message utils
 - Any componenent can include message_utils.h to leverage the same
 - Use message_utils in webrtc_bridge and register for network_coprocessor messages

vikramdattu committed:
Increase robustness of LwsApiCall implementation

 - Handle partial writes by sending data in multiple iterations
 - Use retries when message send fails
 - Track and handle message receive if in parts using `receiveMessage` var

vikramdattu committed:
WebRTC bridge component

 - This component is specifically designed to exchange messages between signaling and streaming planes of the WebRTC
 - For example, if we want to implement only the streaming part of the webrtc and
get signaling messages relayed to it, another device can implement signaling and relay those messages to streaming plane

vikramdattu committed:
Added media_stream component which would be used to record/play audio_video

vikramdattu committed:
Created esp_port directory to contain ESP-IDF ported components and examples

vikramdattu committed:
Added esp_hosted, esp_wifi_remote and network_coprocessor components

 - esp_hosted + esp_wifi_remote enabled chipsets not having wifi/BT to use it via n/w adapter
 - network_coprocessor component can be use with network_adapter

vikramdattu committed:
Added signaling_serializer component

 - It helps packing and unpacking of signaling messages
 - These serialized messages then can be transfered on any transport

vikramdattu committed:
Added documentation for ESP port

 - esp_port/README.md: Added complete description and setup guide for ESP chipsets
 - esp_port/README.md: Describe two solutions: webrtc_classic and split mode
 - Added puml diagrams under esp_port/docs showcasing the workflow in both modes
 - Added a README under example/app_common/spiffs/certs/

vikramdattu commented:
@sirknightj @unicornss do you have comments on this?
If you look at my solution, I have added a `esp_port` directory containing extra needed components and examples specific to ESP-IDF build system. (There are a few, very few, changes in the common source files specific to ESP-IDF under macro checks)
I would like to know if you want to suggest adding a generic port layer which can be leveraged by other platforms to build solutions based on this SDK? For example, this might have `port` directory and different platform ports can stay in the directory etc.

I would like to speed up the porting process as per your comments, so that the repo can then be used in-time for ESP-IDF based open-source solutions, viz., esp-rainmaker and esp-matter etc.

vikramdattu committed:
Added IDF patches for lwip split functioning

 - Patch1: Support LWIP in split_mode. Allows port filtering in lwip_opts.h
 - Patch2: Work-around for STATUS define. KVS too uses `STATUS` for return values. Guard the IDF STATUS under `#ifndef STATUS`

aggarg commented:
You may want to review the reference integration for the Realtek Ameba Pro2 Mini board at: https://github.com/awslabs/freertos-webrtc-reference-on-amebapro-for-amazon-kinesis-video-streams.

This reference integration utilizes our OS-agnostic libraries. You can adopt a similar approach to develop an integration for the ESP platform.

vikramdattu committed:
Added esp32_p4_Function_ev_board component

sirknightj commented on esp_port/README.md:
````diff
@@ -0,0 +1,203 @@
+### Install pkg-config
+
+```bash
+sudo apt-get install pkg-config
````
Would recommend adding the `sudo apt-get update` step just in case.

sirknightj commented on esp_port/README.md:
```diff
@@ -0,0 +1,203 @@
+  - AWS_KVS_LOG_LEVEL
+
+- You may override the above settings in `main/app_main.c` file of the example also.
+
```
Suggest also adding the commands to use the idf.py to:
* Connect the board to your computer (micro usb)
* Build the project
* Flash project onto the ESP board (and any buttons that need to be pressed or held down during the flash, any LED indicators that we can look for)

sirknightj commented on esp_port/README.md:
```diff
@@ -0,0 +1,203 @@
+
+  - ESP_WIFI_SSID
+  - ESP_WIFI_PASSWORD
+  - ESP_MAXIMUM_RETRY
```
I think the majority of the fields here are self explanatory, except ESP_MAXIMUM_RETRY.

Perhaps we can make a table with the options along with a description.

sirknightj commented on esp_port/docs/camera_device.puml:
```diff
@@ -0,0 +1,92 @@
+@startuml
```
I saw this which you might want to look at -- to get these displayed in one of the readmes: https://github.com/Zingam/UML-in-Markdown

Alternatively, GitHub supports mermaid diagrams - https://github.blog/developer-skills/github/include-diagrams-markdown-files-mermaid/

sirknightj commented on esp_port/README.md:
```diff
@@ -0,0 +1,203 @@
+  - AWS_SECRET_ACCESS_KEY
+  - AWS_DEFAULT_REGION
+  - AWS_KVS_CHANNEL
+  - AWS_KVS_LOG_LEVEL
```
I saw some sample frames in one of the folders, would it make sense to have an option to use the camera sensor vs. pre-generated frames?

vikramdattu committed:
Use config option for setting max concurrent streaming sessions

vikramdattu committed:
Refactor: Move logging into separate WebRtcLogging.c

vikramdattu committed:
Refactor(app_webrtc): Move credential provider from webrtcAppInit to webrtcAppRun

 - This allows us keep webrtcAppInit lighter and move webrtcAppRun to different task

vikramdattu committed:
webrtc_classic: remove unused embedded files

vikramdattu committed:
Create a new task for webrtc handling and unblock the caller

vikramdattu committed:
Restruct: Major changes for media_stream, should go in app_webrtc and examples as well

vikramdattu committed:
LwsApiCalls.c: Handling of time_skew

vikramdattu committed:
webrtc_split: move SignalingMessage logic to the app_webrtc.c

 - Application should not need to take care of fitting the message received from webrtc_bridge into SDK's structure format
 - The message will simply be deserialized and passed to SDK using `webrtcAppSignalingMessageReceived` API
 - Messages from SDK are received via callback and we can send to webrtc_bridge from application
 - hosted: Fixed build for IDF release/v5.5

vikramdattu committed:
webrtc_split: move SignalingMessage logic to the app_webrtc.c

 - Application should not need to take care of fitting the message received from webrtc_bridge into SDK's structure format
 - The message will simply be deserialized and passed to SDK using `webrtcAppSignalingMessageReceived` API
 - Messages from SDK are received via callback and we can send to webrtc_bridge from application

vikramdattu committed:
Restruct: defined a default config

 - e.g., config structure can be inited with most common config and fine tuned thereon
```c
    WebRtcAppConfig webrtcConfig = WEBRTC_APP_CONFIG_DEFAULT();
```
 - Bugfix: make a copy of `awsCredentialOptions` so that it can be re-used out of scope

vikramdattu committed:
Leave Wi-Fi configuration for ESP32-C6 to handle

vikramdattu committed:
Move app_webrtc.c/.h to kvs_webrtc component

vikramdattu committed:
Updated lwip port filtering patch

vikramdattu committed:
Updated usersctp to the threadsafe version

vikramdattu committed:
esp_work_queue: New API to init workq with config

vikramdattu committed:
Improve flash_wrapper functionality to handle NVS ops disguised as fileio

 - This is helpful where our cert files are actually in NVS
 - Also updated the README for esp_webrtc_utils component
 - app_webrtc.c: Use fileIO wrapper functions instead of direct file ops

vikramdattu committed:
TimeSkew (2/n): remove SNTP time sync dependency for esp_websocket_client

 - We simply store the server time globally when we do http requests
 - Use that gloabal time for aws signature and Date header while websocket connect

vikramdattu committed:
esp_hosted: Fix BT functionality

 - Can be useful for provisioning workflows OR any other use cases where host needs BT

vikramdattu committed:
esp_hosted: Modify init timer timeout 1s --> 10s

 - Aggressive timeout makes board reboot prematurely.
 - Wait for max 10 seconds before SDIO transport is ready

vikramdattu committed:
app_webrtc: Added and API to trigger offer

 - New API `webrtcAppCreateAndSendOffer`, can be called to initiate the offer
 - Also, spin the mediaSender thread in handlerAnswer

vikramdattu committed:
app_webrtc: Added and API to trigger offer

 - New API `webrtcAppCreateAndSendOffer`, can be called to initiate the offer
 - Also, spin the mediaSender thread in handlerAnswer

vikramdattu committed:
Example show-casing apprtc based esp camera

 - Works in Viewer mode of KVS WebRTC
 - The camera registers to a room and initiates the sdp offer
 - The client can join the room(this sends the sdp answer to camera)
 - peer connection is established and feed can be viewed

vikramdattu committed:
Restructure app_webrtc.c to move out signaling to separate file

vikramdattu committed:
Continue with the peer connection even if getIceServerConfig failed

 - TURN servers are not `MUST` for the peer connection
 - The device can still continue with the STUN server
 - In cases where, server is configured to not allow IceServers, we can continue

vikramdattu committed:
Improve connection robustness

 - Attempt reconnections from webrtc task when disconnections happen
   - Reconnections are handled with new task in LwsApiCalls.c
   - Workflow was broken with LwsApiCallsESP.c: We now handle it from webrtc_task itself
 - Do not reboot the device when using wifi-set CLI command

vikramdattu commented:
Closing this PR — it has been superseded. The ESP-IDF port now lives in two better homes:
- the [`beta-reference-esp-port`](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/tree/beta-reference-esp-port) branch in this repo, which carries a reference copy of the port, and
- the dedicated port repository, [espressif/esp-port-for-amazon-kvs-sdk](https://github.com/espressif/esp-port-for-amazon-kvs-sdk), where active development happens (examples, CI, board support).

Thanks to everyone who reviewed here — the feedback shaped the port as it exists today.
